app/org/sagebionetworks/bridge/services/SessionUpdateService.java

```diff
@@ -0,0 +1,147 @@
+
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.joda.time.DateTimeZone;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import org.sagebionetworks.bridge.cache.CacheProvider;
+import org.sagebionetworks.bridge.dao.ParticipantOption.SharingScope;
+import org.sagebionetworks.bridge.models.CriteriaContext;
+import org.sagebionetworks.bridge.models.accounts.ConsentStatus;
+import org.sagebionetworks.bridge.models.accounts.ExternalIdentifier;
+import org.sagebionetworks.bridge.models.accounts.StudyParticipant;
+import org.sagebionetworks.bridge.models.accounts.UserSession;
+import org.sagebionetworks.bridge.models.subpopulations.SubpopulationGuid;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Service that updates the state of a user's session, eventually its dependencies as well (the state of the 
+ * user's push notification topic subscriptions, and the session as it is cached). Changes that update a 
+ * user's session should go through this service to ensure dependencies are handled correctly.
+ */
+@Component
+public class SessionUpdateService {
+    
+    private CacheProvider cacheProvider;
+    
+    private ConsentService consentService;
+    
+    @Autowired
+    public final void setCacheProvider(CacheProvider cacheProvider) {
+        this.cacheProvider = cacheProvider;
+    }
+    @Autowired
+    public final void setConsentService(ConsentService consentService) {
+        this.consentService = consentService;
+    }
+    
+    public void updateTimeZone(UserSession session, DateTimeZone timeZone) {
+        session.setParticipant(builder(session).withTimeZone(timeZone).build());
+        cacheProvider.setUserSession(session);
+    }
+    
+    public void updateLanguage(UserSession session, LinkedHashSet<String> languages) {
+        session.setParticipant(builder(session).withLanguages(languages).build());
+        cacheProvider.setUserSession(session);
+    }
+    
+    public void updateExternalId(UserSession session, ExternalIdentifier externalId) {
+        session.setParticipant(builder(session).withExternalId(externalId.getIdentifier()).build());
+        cacheProvider.setUserSession(session);
+    }
+    
+    public void updateParticipant(UserSession session, StudyParticipant participant) {
+        session.setParticipant(participant);
+        cacheProvider.setUserSession(session);
+    }
+    
+    public void updateParticipant(UserSession session, CriteriaContext context, StudyParticipant participant) {
+        session.setParticipant(participant);
+        
+        Map<SubpopulationGuid,ConsentStatus> statuses = consentService.getConsentStatuses(context);
+        session.setConsentStatuses(statuses);
+        
+        cacheProvider.setUserSession(session);
+    }
+    
+    public void updateDataGroups(UserSession session, CriteriaContext context, Set<String> dataGroups) {
```

Comment:
I'm confused by this method signature (and updateParticipant above) and how it should be called.

CriteriaContext also contains the participant's data groups. Does the dataGroups param need to match the data groups in the context param? 

Looks to me like the steps are
- get new dataGroups
- set new dataGroups on context
- call updateDataGroups with params: old session, and updated context and dataGroup
- the old session passed in as param gets updated with data groups and consents as a side effect

Is that right?